macaroonbakery/bakery.py

```diff
@@ -0,0 +1,280 @@
+            as a base64 encoded json (dict) macaroon.
+        @param timeout timeout for wait url.
+        '''
+        self._macaroons = initial_macaroons
+        self._discharge_token = discharge_token
+        self._timeout = timeout
+
+    def get(self, url, **kwargs):
+        '''Act as the same than requests.get, same arguments,
+           same returned value, same exception raised except DischargeError
+           when something wrong happened when discharging.
+           If macaroons are present, the headers will be augmented with it.
+        '''
+        if self._macaroons is not None:
+            headers = kwargs.setdefault('headers', {})
+            headers.update({
+                'Bakery-Protocol-Version': 1,
+                'Macaroons': self._macaroons,
+            })
+        response = requests.get(url, **kwargs)
+        if (response.status_code == 401 and
+                response.headers['Www-Authenticate'] == 'Macaroon') \
+                or response.status_code == 407:
+            try:
+                response_json = response.json()
+            except ValueError:
+                raise DischargeError(
+                    'Unable to get data from unauthorized response')
+            try:
+                code = response_json['Code']
+            except KeyError:
+                raise DischargeError(
+                    'code not found in unauthorized response {}'.format(
+                        response.content))
+            if code != 'macaroon discharge   required':
+                raise DischargeError(
+                    'Unknown code found in unauthorized response {}'.format(
+                        code))
+            try:
+                serialized_macaroon = response_json['Info']['Macaroon']
+            except KeyError:
+                raise DischargeError(
+                    'macaroon not found in unauthorized response {}'.format(
+                        response.content))
+            macaroon = deserialize(serialized_macaroon)
+            discharges = self._discharge(macaroon)
+            encoded_discharges = map(serialize_macaroon_string, discharges)
+            self._macaroons = base64.urlsafe_b64encode(
+                '[' + ','.join(encoded_discharges) + ']')
+            return self.get(url, **kwargs)
+
+        return response
+
+    def _discharge(self, macaroon):
+        '''Discharge a macaroon.
+
+        @param the macaroon object to discharge.
+        @returns the discharged macaroons as an array of macaroon object.
+        @raise DischargeError when an error occurs during the discharge
+            process.
+        '''
+        discharges = [macaroon]
+        first_party_location = macaroon.location
+        self._discharge_caveats(macaroon, discharges, first_party_location,
+                                macaroon)
+        return discharges
+
+    def _discharge_caveats(self, macaroon, discharges, first_party_location,
```

Comment:
FWIW we've deprecated the first party location and it's never actually been used, so perhaps don't bother passing it to the discharger.